File: app/src/test/java/com/coinninja/coinkeeper/cn/wallet/HDWalletTest.java
package com.coinninja.coinkeeper.cn.wallet;

import com.coinninja.bindings.DecryptionKeys;
import com.coinninja.bindings.DerivationPath;
import com.coinninja.bindings.EncryptionKeys;
import com.coinninja.bindings.Libbitcoin;
import com.coinninja.coinkeeper.cn.wallet.data.HDWalletTestData;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class HDWalletTest {

    @Mock
    LibBitcoinProvider libBitcoinProvider;

    @Mock
    private Libbitcoin libbitcoin;

    @InjectMocks
    private HDWallet wallet;

    @Before
    public void setUp() throws Exception {
        when(libBitcoinProvider.provide()).thenReturn(libbitcoin);
        prepareAddresses();
    }

    @Test
    public void fillsExternalBlocksAtStartingIndex() {
        String[] addresses = wallet.fillBlock(HDWallet.EXTERNAL, 0, 50);

        assertThat(addresses, equalTo(HDWalletTestData.blockOneExternalAddresses));
    }

    @Test
    public void fillsInternalBlocksAtStartingIndex() {
        String[] addresses = wallet.fillBlock(HDWallet.INTERNAL, 0, 50);


        assertThat(addresses, equalTo(HDWalletTestData.blockOneInternalAddresses));
    }

    @Test
    public void returns_external_address_for_index() {
        String address = "--ex-address";
        when(libbitcoin.getExternalChangeAddress(0)).thenReturn(address);

        assertThat(wallet.getExternalAddress(0), equalTo(address));
    }

    @Test
    public void generateEncryptionKeys() {
        String publicKey = "public";
        EncryptionKeys encryptionKeys = new EncryptionKeys("".getBytes(), "".getBytes(), "".getBytes());
        when(libbitcoin.getEncryptionKeys(publicKey)).thenReturn(encryptionKeys);

        EncryptionKeys keys = wallet.generateEncryptionKeys(publicKey);
        assertThat(keys, equalTo(encryptionKeys));
    }

    @Test
    public void generateDecryptionKeys() {
        DecryptionKeys decryptionKeys = new DecryptionKeys("".getBytes(), "".getBytes());

        DerivationPath derivationPath = new DerivationPath("m/49/0/0/0/1");

        byte[] ephemeralPublicKey = "ephemeralPublicKey".getBytes();
        when(libbitcoin.getDecryptionKeys(derivationPath, ephemeralPublicKey)).thenReturn(decryptionKeys);

        DecryptionKeys keys = wallet.generateDecryptionKeys(derivationPath, ephemeralPublicKey);

        assertThat(keys, equalTo(decryptionKeys));
    }

    private void prepareAddresses() {
        //External
        for (int i = 0; i < HDWalletTestData.blockOneExternalAddresses.length; i++) {
            when(libbitcoin.getExternalChangeAddress(i)).thenReturn(HDWalletTestData.blockOneExternalAddresses[i]);
        }

        //Internal
        for (int i = 0; i < HDWalletTestData.blockOneExternalAddresses.length; i++) {
            when(libbitcoin.getInternalChangeAddress(i)).thenReturn(HDWalletTestData.blockOneInternalAddresses[i]);
        }
    }
}

File: app/src/test/java/com/coinninja/coinkeeper/cn/wallet/HDWalletTest.kt
package com.coinninja.coinkeeper.cn.wallet

import com.coinninja.bindings.DecryptionKeys
import com.coinninja.bindings.DerivationPath
import com.coinninja.bindings.EncryptionKeys
import com.coinninja.bindings.Libbitcoin
import com.coinninja.coinkeeper.cn.wallet.data.HDWalletTestData
import com.coinninja.coinkeeper.model.db.Wallet
import com.google.common.truth.Truth.assertThat
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import org.junit.Test

class HDWalletTest {

    private val libbitcoin: Libbitcoin = mock()
    private val wallet: Wallet = mock()

    private fun createHDWwallet(): HDWallet {
        val hdWallet = HDWallet(mock(), mock())
        whenever(hdWallet.libBitcoinProvider.provide()).thenReturn(libbitcoin)
        whenever(hdWallet.walletHelper.wallet).thenReturn(wallet)
        whenever(wallet.purpose).thenReturn(49)
        whenever(wallet.coinType).thenReturn(1)
        whenever(wallet.accountIndex).thenReturn(0)
        return hdWallet
    }

    @Test
    fun fillsExternalBlocksAtStartingIndex_external() {
        val wallet = createHDWwallet()
        val bufferSize = 50
        prepareAddresses(HDWallet.EXTERNAL, bufferSize)
        val addresses = wallet.fillBlock(HDWallet.EXTERNAL, 0, bufferSize)

        assertThat(addresses).isEqualTo(HDWalletTestData.blockOneExternalAddresses)
    }

    @Test
    fun fillsInternalBlocksAtStartingIndex_internal() {
        val wallet = createHDWwallet()
        val bufferSize = 50
        prepareAddresses(HDWallet.INTERNAL, bufferSize)
        val addresses = wallet.fillBlock(HDWallet.INTERNAL, 0, bufferSize)


        assertThat(addresses).isEqualTo(HDWalletTestData.blockOneInternalAddresses)
    }

    @Test
    fun returns_address_for_derivation_path() {
        val wallet = createHDWwallet()
        val address = "--ex-address"
        val derivationPath = DerivationPath(49, 0, 0, 0, 0)
        whenever(libbitcoin.getAddressForPath(derivationPath)).thenReturn(address)

        assertThat(wallet.getAddressForPath(derivationPath)).isEqualTo(address)
    }

    @Test
    fun generateEncryptionKeys() {
        val wallet = createHDWwallet()
        val publicKey = "public"
        val encryptionKeys = EncryptionKeys("".toByteArray(), "".toByteArray(), "".toByteArray())
        whenever(libbitcoin.getEncryptionKeys(publicKey)).thenReturn(encryptionKeys)

        val keys = wallet.generateEncryptionKeys(publicKey)
        assertThat(keys).isEqualTo(encryptionKeys)
    }

    @Test
    fun generateDecryptionKeys() {
        val wallet = createHDWwallet()
        val decryptionKeys = DecryptionKeys("".toByteArray(), "".toByteArray())
        val derivationPath = DerivationPath("m/49/0/0/0/1")
        val ephemeralPublicKey = "ephemeralPublicKey".toByteArray()

        whenever(libbitcoin.getDecryptionKeys(derivationPath, ephemeralPublicKey)).thenReturn(decryptionKeys)

        val keys = wallet.generateDecryptionKeys(derivationPath, ephemeralPublicKey)

        assertThat(keys).isEqualTo(decryptionKeys)
    }

    private fun prepareAddresses(chainIndex: Int, bufferSize: Int) {
        if (chainIndex == HDWallet.EXTERNAL) {
            for (i in 0 until bufferSize) {
                whenever(libbitcoin.getAddressForPath(DerivationPath(wallet.purpose, wallet.coinType, wallet.accountIndex, HDWallet.EXTERNAL, i))).thenReturn(HDWalletTestData.blockOneExternalAddresses[i])
            }
        } else if (chainIndex == HDWallet.INTERNAL) {
            for (i in 0 until bufferSize) {
                whenever(libbitcoin.getAddressForPath(DerivationPath(wallet.purpose, wallet.coinType, wallet.accountIndex, HDWallet.INTERNAL, i))).thenReturn(HDWalletTestData.blockOneInternalAddresses[i])
            }

        }
    }
}